Src/AutoFixture/MailAddressGenerator.cs

```diff
@@ -36,10 +36,21 @@ public object Create(object request, ISpecimenContext context)
                 return new NoSpecimen(request);
             }
 
-            var name = Guid.NewGuid().ToString();
-            var host = fictitiousDomains[(uint)name.GetHashCode() % 3];
+            if (context == null)
+            {
+                return new NoSpecimen(request);

```

Comment:
This is perhaps a little confusing, but while `request` is a bit special when it comes to null handling, `context` follows the [standard *design-by-contract* principles laid out by Bertrand Meyer](http://amzn.to/1claOin):

A method's *preconditions* describe what the method *needs* in order to do its work. Its *postconditions* provide a guarantee about the state of the system once the method returns - if the precondition was satisfied.

In this case, the postcondition should be that a `MailAddress` should be returned if the preconditions are met. In order to create a `MailAddress`, the new implementation *needs* the `context`. Thus, if the `context` is null, the precondition isn't satisfied.

That's an error.

Returning `NoSpecimen` means that the builder doesn't *want* to handle that particular `request`; it doesn't mean that it *can't*.

As I'm trying to explain this, I realize that this seems rather subtle, but that's the way the rest of AutoFixture works, so if the `context` is null, it should throw an `ArgumentNullException`.